jobs/kpi-forecasting/kpi_forecasting/validator.py

```diff
@@ -0,0 +1,153 @@
+                    dimensions, where different combinations of values specify separate models
+                    If a model has no such columns, set to an empty list
+
+        Raises:
+            Exception: Raised if list of config files have different values for the dimension list
+            Exception: Raised if list of config files have different values for the input table
+        """
+        segment_data_list = []
+        input_table_list = []
+        config_file_list = list(self.config_data.keys())
+        for config_data in self.config_data.values():
+            # get segment data
+            metric_hub_data = config_data.metric_hub.toDict()
+            if "segments" in metric_hub_data:
+                segment_data = metric_hub_data["segments"]
+                segment_data_list.append(segment_data)
+            else:
+                segment_data_list.append(None)
+
+            # get input table info
+            input_table_list.append(config_data.write_results.toDict())
+
+        input_table_data = input_table_list.pop(0)
+        input_table_matches_first = [input_table_data == el for el in input_table_list]
+        if not all(input_table_matches_first):
+            config_file_list_string = " ".join(config_file_list)
+            raise Exception(
+                f"Input Table Data Does not all match for config list: {config_file_list_string}"
+            )
+
+        input_project = input_table_data["project"]
+        input_dataset = input_table_data["dataset"]
+        input_table = input_table_data["table"]
+
+        input_table_full = f"{input_project}.{input_dataset}.{input_table}"
+
+        segment_data = segment_data_list.pop(0)
+        segment_data_matches_first = [segment_data == el for el in segment_data_list]
+        if not all(segment_data_matches_first):
+            config_file_list_string = " ".join(config_file_list)
+            raise Exception(
+                f"Dimension Data Does not all match for config list: {config_file_list_string}"
+            )
+
+        if segment_data:
+            # this is the case where dimensions are present
+            # we only need the column names for the query
+            dimension_list = list(segment_data.keys())
+        else:
+            dimension_list = []
+
+        self.input_table_full = input_table_full
+        self.dimension_list = dimension_list
+
+    def query(self) -> str:
+        """Creates a query used to generate the validation table.
+
+        Returns:
+        (str): Query to generate validation table"""
+        if len(self.dimension_list) > 0:
+            segment_comma_separated = "," + ",".join(self.dimension_list)
+        else:
+            segment_comma_separated = ""
+        # in actual_deduped, the value for historical data won't change so we can use any_value without checking forecast_trained_at
+        compare_ctes = f"""WITH forecast_deduped as (SELECT submission_date, metric_alias {segment_comma_separated},
```

Comment:
I think we should pull this out into its own class. It'll be necessary to use a similar query for adjusting the forecasts, like for product change impacts, so it should be somewhere accessible outside of the `Validator` class.

Reply:
Although you are doing all of the metric calculation in the SQL. Are you opposed to making a general class to pull the data and using Python to calculate errors?

Reply:
Not opposed to doing the calculations in python.  Could you explain the other use case for the pull, so I can think through how to generalize it?

Reply:
The other case, currently, would be to pull forecast data to make post-hoc adjustments -- for instance, when product thinks a change will impact the forecasts by x amount, pull the forecasts and apply that x scalar.